### PR TITLE
Add RepositoryVersion Filter

### PR DESCRIPTION
**Pull request by werwty**
fixes #3238
https://pulp.plan.io/issues/3238

### CODE REVIEW

**Comment by werwty** [comment]
The filter can be tested with the following query: http://pulp3.dev:8000/api/v3/repositories/f6bf16e4-e419-482b-8032-942e3fa5b93b/versions/?version_range=3,4&created_0=2018-02-06 19:27&created_1=&content_in=1ac31f86-04ff-47cf-a3ff-2578dcaebeec,00ab453f-dbb0-481e-9934-cf68784989dd

**Comment by daviddavis** [comment]
I was thinking that the ranges would be like two fields like `created_after` and `created_before`. Not sure I understand the field names `created_0` and `created_1`.

Also, I think that the content filter should accept hrefs like we do when you create a repo version with a list of content units:

https://github.com/pulp/pulp/blob/3.0-dev/pulpcore/pulpcore/app/viewsets/repository.py#L227

**Comment by daviddavis** [comment]
It's late so apologies if I am not thinking clearly but it seems like your content filter algorithm is not removing one version number when it generates it range. In other words, if I search for a content unit that was added in version 4 and removed in version 6, will your code with search versions 4-6 or will it only search versions 4 and 5?

**Comment by daviddavis** [comment]
Nevermind, I see that `range(4, 6)` will return `[4, 5]`. Ignore my previous comment 😄

**Comment by daviddavis** [comment]
Is this `version_range` really a `range` filter or is it a `in` filter? I would expect `version_range=3,6` to return all versions from 3-6 whereas with `version_in=3,6` I would expect versions 3 and 6 to be returned.

**Comment by werwty** [comment]
Thanks for the review! I updated so that instead of trying to use a fancy DateTimeFromToRangeFilter (which generated the `created_0` and `created_1` queries, that needed a custom widget to override https://django-filter.readthedocs.io/en/master/ref/widgets.html#suffixedmultiwidget ) I just separated it into two filters.

I did the same for the VersionRange filter, since a `NumberRangeFilter(BaseRangeFilter, NumberFilter)` expected both a lower and upper bound, and doing it through two filters does not.

`RepositoryVersionContentInFilter` was also updated to take content hrefs.

File: pulpcore/pulpcore/app/viewsets/repository.py [hunk 1]
```diff
@@ -128,6 +129,86 @@ class Meta:
...
+        2. Get a list of version_added and version_removed where the content was
+           changed on the repository
+        3. Calculate and return the versions that the contents can be found on
+
+    """
+
+    def filter(self, qs, value):
+        """
+        Args:
+            qs (pulpcore.app.viewsets.repository.RepositoryVersionViewSet.queryset):
```
Comment (asmacdo) [inline pulpcore/pulpcore/app/viewsets/repository.py]: This type is very informative, but I think it violates the spirit of the docstring conventions. It probably ought to be:

qs (django.db.models.query.QuerySet)

File: pulpcore/pulpcore/app/viewsets/repository.py [hunk 2]
```diff
@@ -128,6 +129,86 @@ class Meta:
...
+class RepositoryVersionContentInFilter(CharInFilter):
+    """
+    Filter used to get the repository versions where some given content can be found.
+
+    Given a list of content_id, this filter will:
+        1. Get the RepositoryContent that the id can be found in
+        2. Get a list of version_added and version_removed where the content was
+           changed on the repository
+        3. Calculate and return the versions that the contents can be found on
+
```
Comment (asmacdo) [inline pulpcore/pulpcore/app/viewsets/repository.py]: No space here

File: pulpcore/pulpcore/app/viewsets/repository.py [hunk 3]
```diff
@@ -30,17 +32,16 @@
...
                                    GenericNamedModelViewSet,
                                    CreateReadAsyncUpdateDestroyNamedModelViewset)
-from pulpcore.app.viewsets.custom_filters import CharInFilter
+from pulpcore.app.viewsets.custom_filters import CharInFilter, NumberRangeFilter
 from pulpcore.common import tags
 
 
 class RepositoryFilter(filterset.FilterSet):
     name_in_list = CharInFilter(name='name', lookup_expr='in')
-    content_added_since = filters.Filter(name='last_content_added', lookup_expr='gt')
```
[the remark this answers is not in the document]
Comment (werwty) [reply pulpcore/pulpcore/app/viewsets/repository.py]: the last_content_added field was removed from Repository a while back

**Comment by daviddavis** [comment]
The code looks good. I just realized though: if I filter on content_in=A,B,C and I have a version that contains only A, will it be returned?

**Comment by werwty** [comment]
@daviddavis it would be returned. Currently this is an 'OR' filter, and would return all repositories containing (A | B | C)

File: pulpcore/pulpcore/app/viewsets/repository.py [hunk 4]
```diff
@@ -128,6 +129,87 @@ class Meta:
...
+
+        Returns:
+            Queryset of the RepositoryVersions containing specified content
+        """
+
+        # If no value is passed, return queryset as-is
+        if not value:
+            return qs
+
+        contend_pks = [RepositoryVersionViewSet.get_content_unit_pk(url) for url in value]
```
Comment (daviddavis) [inline pulpcore/pulpcore/app/viewsets/repository.py]: [A PR just got merged that affects this](https://github.com/pulp/pulp/pull/3311). I think you just need this now:

```python
contend_pks = [GenericNamedModelViewSet.get_resource(url, Content).pk for url in value]
```

**Comment by daviddavis** [comment]
One comment. Otherwise LGTM.

File: pulpcore/pulpcore/app/viewsets/custom_filters.py [hunk 5]
```diff
@@ -9,3 +9,8 @@ class CharInFilter(filters.BaseInFilter, filters.CharFilter):
     Enables the user to filter a field by comma separated strings, allowing them to retrieve more
     than one object in a single query.
     """
+    pass
+
+
+class NumberRangeFilter(filters.BaseInFilter, filters.NumberFilter):
```
[the remark this answers is not in the document]
Comment (werwty) [reply pulpcore/pulpcore/app/viewsets/custom_filters.py]: I updated this to use 2 NumberFilters:
```
version_min = filters.NumberFilter(name='number', lookup_expr='gte')
version_max = filters.NumberFilter(name='number', lookup_expr='lte')```

Mainly so we don't get `version_0=&version_1=` in the field name

**Pull request merged**